cilium: populate wildcard src->dst policy for transparent encryption

jrfastab opened the pull request:
Policy/State rules are populated when we receive kvstore or k8s
events. On egress we must match the (dst_addr, spi) tuple via the xfrm
policy and then link this to a state using the tmpl including the
same (dst_addr, spi). This results in a pair of policy/state rules
per node on the host. The above rules also specify a src address
so that we only match Cilium managed traffic. This includes the
IP address on cilium_host and the endpoint allocation range.

On the ingress side we have been doing the same but swapped. The
tuple (src, dst, spi) is used in the policy to match traffic and
then linked to the state via the tmpl. These are also added on
update events. However, this creates more rule than are actually
needed. By wildcarding the source on ingress traffic we can
match all traffic for the node with a single rule. Further the
existing scheme is buggy and dst-ip/spi pairs can collide.

The fix is to insert a wildcard decrypt rule when the local node
update event is received. Then on node update events only
add the egress rule.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7181)
<!-- Reviewable:end -->

jrfastab committed:
cilium: populate wildcard src->dst policy for ipsec

Policy/State rules are populated when we receive kvstore or k8s
events. On egress we must match the (dst_addr, spi) tuple via the xfrm
policy and then link this to a state using the tmpl including the
same (dst_addr, spi). This results in a pair of policy/state rules
per node on the host. The above rules also specify a src address
so that we only match Cilium managed traffic. This includes the
IP address on cilium_host and the endpoint allocation range.

On the ingress side we have been doing the same but swapped. The
tuple (src, dst, spi) is used in the policy to match traffic and
then linked to the state via the tmpl. These are also added on
update events. However, this creates more rule than are actually
needed. By wildcarding the source on ingress traffic we can
match all traffic for the node with a single rule. Further the
existing scheme is buggy and dst-ip/spi pairs can collide.

The fix is to insert a wildcard decrypt rule when the local node
update event is received. Then on node update events only
add the egress rule.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

aanm commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -33,6 +33,14 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+type IPSecDir string
```
For a moment I thought this was the "IPSecDirectory" and not "...Direction"

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -429,12 +429,18 @@ func (n *linuxNodeHandler) enableIPsec(newNode *node.Node) {
+			ciliumInternalIPv4 := newNode.GetCiliumInternalIP(false)
+			if ciliumInternalIPv4 != nil {
+				ipsecLocal := &net.IPNet{IP: n.nodeAddressing.IPv4().Router(), Mask: n.nodeAddressing.IPv4().AllocationCIDR().Mask}
+				ipsecIPv4Wildcard := &net.IPNet{IP: net.ParseIP("0.0.0.0"), Mask: net.IPv4Mask(0, 0, 0, 0)}
```
small nit, consider making this a "static" / "constant" variable in this file

jrfastab replied:
will do.

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -449,15 +455,21 @@ func (n *linuxNodeHandler) enableIPsec(newNode *node.Node) {
+			ciliumInternalIPv6 := newNode.GetCiliumInternalIP(true)
+			if ciliumInternalIPv6 != nil {
+				ipsecLocal := &net.IPNet{IP: n.nodeAddressing.IPv6().Router(), Mask: n.nodeAddressing.IPv6().AllocationCIDR().Mask}
+				ipsecIPv6Wildcard := &net.IPNet{IP: net.ParseIP("0::0"), Mask: net.CIDRMask(0, 0)}
```
small nit, consider making this a "static" / "constant" variable in this file

jrfastab replied:
will do

aanm reviewed (commented):
nothing blocking on my side

jrfastab committed:
cilium: push decryption up so we can decrypt even if not endpoint

Currently we check if there is a corresponding endpoint for the ip4
header before decrypting it. But, if we are encrypting the traffic
to the tunnel for an unmanaged pod we mey not have an endpoint in
the table. If this is the case move ESP check out of if/else and
decrypt using IPsec stack.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>
